src/Core/Resolvers/SqlMutationEngine.cs

```diff
@@ -367,135 +376,383 @@ public class SqlMutationEngine : IMutationEngine
+                // It returns false if the primary key does not exist.
+                if (sqlMetadataProvider.TryGetExposedColumnName(context.EntityName, primarykey, out string? exposedColumnName))
+                {
+                    findRequestContext.PrimaryKeyValuePairs.Add(exposedColumnName, mutationResultRow.Columns[exposedColumnName]!);
+                }
+                else
+                {
+                    // This code block should never be reached because the information about primary keys gets populated during the startup.
+                    throw new DataApiBuilderException(
+                       message: "Insert/Upsert operation was successful but unexpected error when constructing the response",
+                       statusCode: HttpStatusCode.InternalServerError,
+                       subStatusCode: DataApiBuilderException.SubStatusCodes.UnexpectedErrorFindingExposedFieldName);
+                }
+            }
+
+            // READ action for the given role can have include and exclude fields configured. Populating UpdateReturnFields
+            // ensures that the select query retrieves only those fields that are allowed for the given role.
+            findRequestContext.UpdateReturnFields(_authorizationResolver.GetAllowedExposedColumns(context.EntityName, roleName, EntityActionOperation.Read));
+
+            return findRequestContext;
+        }
+
+        /// <summary>
+        /// Constructs and returns a HTTP 201 Created response.
+        /// The response is constructed using results of the upsert database operation when database policy is not defined for the read permission.
+        /// If database policy is defined, the results of the subsequent select statement is used for constructing the response.
+        /// </summary>
+        /// <param name="resultRow">Reuslt of the upsert database operation</param>
+        /// <param name="jsonDocument">Result of the select database operation</param>
+        /// <param name="primaryKeyRoute">Primary key route to be used in the Location Header</param>
+        /// <param name="isReadPermissionConfiguredForRole">Indicates whether read permissions is configured for the role</param>
+        /// <param name="isDatabasePolicyDefinedForReadAction">Indicates whether database policy is configured for read action</param>
+        /// <param name="httpContext">HTTP Context associated with the API request.</param>
+        /// <param name="operationType">Resultant Operation type - Update, Insert, etc.</param>
+        private CreatedResult ConstructCreatedResultResponse(
+            Dictionary<string, object?> resultRow,
+            JsonDocument? jsonDocument,
+            string primaryKeyRoute,
+            bool isReadPermissionConfiguredForRole,
+            bool isDatabasePolicyDefinedForReadAction,
+            EntityActionOperation operationType,
+            HttpContext httpContext)
```

Comment:
Is it necessary to add HttpContext to the method signature? Since httpcontextaccessor is a dependency of the mutationengine, you could just access it here.

Reply:
This was a miss. Updated to use the `_httpContextAccessor`